streams/src/main/java/org/apache/kafka/streams/state/internals/buffer/InMemoryTimeOrderedKeyValueBufferStore.java

```diff
@@ -0,0 +1,219 @@
+        @Override
+        public StoreBuilder<StateStore> withLoggingDisabled() {
+            return null;
+        }
+
+        @Override
+        public StateStore build() {
+            return new InMemoryTimeOrderedKeyValueBufferStore(
+                new InMemoryTimeOrderedKeyValueBuffer(),
+                storeName
+            );
+        }
+
+        @Override
+        public Map<String, String> logConfig() {
+            return Collections.emptyMap();
+        }
+
+        @Override
+        public boolean loggingEnabled() {
+            return true;
+        }
+
+        @Override
+        public String name() {
+            return storeName;
+        }
+    }
+
+    private InMemoryTimeOrderedKeyValueBufferStore(final InMemoryTimeOrderedKeyValueBuffer buffer,
+                                                   final String storeName) {
+        this.buffer = buffer;
+        this.storeName = storeName;
+    }
+
+    public InMemoryTimeOrderedKeyValueBuffer getBuffer() {
+        return buffer;
+    }
+
+    @Override
+    public String name() {
+        return storeName;
+    }
+
+    @Override
+    public void close() {
+        buffer.clear();
+        open = false;
+    }
+
+    @Override
+    public boolean persistent() {
+        return false;
+    }
+
+    @Override
+    public boolean isOpen() {
+        return open;
+    }
+
+    @Override
+    public void init(final ProcessorContext context, final StateStore root) {
+        context.register(root, (RecordBatchingStateRestoreCallback) this::restoreBatch);
+        try {
+            collector = ((RecordCollector.Supplier) context).recordCollector();
+            changelogTopic = ProcessorStateManager.storeChangelogTopic(context.applicationId(), storeName);
+        } catch (final UnsupportedOperationException e) {
+            LOG.warn("Disabling durability, since the context doesn't support it.", e);
+        }
+        open = true;
+    }
+
+    @Override
+    public void flush() {
+        // counting on this getting called before the record collector's flush
+        for (final Bytes key : buffer.getAndClearDirtyKeys()) {
+            final KeyValue<Long, ContextualRecord> bufferTimeAndValue = buffer.get(key);
+            if (bufferTimeAndValue == null) {
+                // The record was evicted from the buffer. Send a tombstone.
+                collector.send(changelogTopic, key, null, null, null, null, KEY_SERIALIZER, VALUE_SERIALIZER);
+            } else {
+                final ContextualRecord value = bufferTimeAndValue.value;
+                final byte[] innerValue = value == null ? null : value.value();
+                final ProcessorRecordContext recordContext = value == null ? null : value.recordContext();
+                final Headers headers = recordContext == null ? new RecordHeaders() : new RecordHeaders(recordContext.headers());
+                headers.add(BUFFER_TIME_HEADER, ByteBuffer.wrap(new byte[8]).putLong(bufferTimeAndValue.key).array());
```

Comment:
The two options are to pack this timestamp in the value or to add it as a header.

I'm not too familiar with the headers... Are there any considerations that would cause us to choose one over the other?

Reply:
Header need to store the headers key; thus, it's more space overhead. Also, as your own method `extractBufferTimeHeader` shows, it might interfere with user headers making the code "ugly". Might be opinionated, but I tend to prefer encoding it in the value instead of using headers (seems to be "cleaner").